pill: Open user info popover on clicking on a user pill

Pranav2612000 opened the pull request:
The user pills on the settings page looked clickable, but clicking them had no effect. This commit fixes that by opening the user info popover on clicking on a user pill in an editable input.

We do this by attaching an event handler to all .pill containers from the body. This handler picks up the pill user's id and renders a popover for the user. For this to work, we also store the user_id in the input_pill template.

Fixes #24247 
<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
![Kapture 2023-02-09 at 08 46 40](https://user-images.githubusercontent.com/20909078/217709094-922cb53a-f215-4602-ac9b-c7106100ee6d.gif)

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

Pranav2612000 commented:
As suggested by @alya 
Here's how it affects user pills in PMs ( clicking a user pill opens up the card ) and search typeaheads ( no effect )

![Kapture 2023-02-10 at 09 37 09](https://user-images.githubusercontent.com/20909078/217999050-ed11f982-3316-4037-a712-76dd272679c7.gif)

sahil839 reviewed (commented):
@Pranav2612000 Thanks for working on this! Posted a couple of initial comments. You would also need to rebase the PR for the next review.

Pranav2612000 commented on static/js/click_handlers.js:
```diff
@@ -410,7 +410,6 @@ export function initialize() {
     });
 
     // RECENT TOPICS
-
```
Fixed.

Pranav2612000 commented on static/js/input_pill.js:
```diff
@@ -98,6 +98,7 @@ export function create(opts) {
             const has_image = item.img_src !== undefined;
 
             const opts = {
+                user_id: item.user_id,
```
Thank you. Will check once.By User groups do you mean the ones in PMs, where the user can add users in the To column in compose box?

sahil839 replied:
You can see this by typing user group name in the UI for adding subscribers to stream in stream settings UI.

Pranav2612000 replied:
Is this what you were talking about?
![image](https://user-images.githubusercontent.com/20909078/220815607-c625dd10-8803-44ee-a7c6-03eefce7072b.png)

It's working as expected. What do you think?

sahil839 replied:
Type a user group name and select it from the typeahead. There is a group named `hamletcharacters`.

Pranav2612000 replied:
Thank you so much @sahil839 for pointing this out. It was a great catch.
I've updated the code to only work for those pills who have a user_id assigned to them. Can you take a look at the code again when you have time?

sahil839 commented on web/src/click_handlers.js:
```diff
@@ -835,6 +836,21 @@ export function initialize() {
+        }
+
+        const user = people.get_by_user_id(participant_user_id);
+        popovers.show_user_info_popover(this, user);
```
I think we should have this in `input_pill.js` where have the click handler for `x` button.

Pranav2612000 replied:
Updated. Thank you @sahil839

sahil839 reviewed (commented):
Posted one comment. Looks good to me otherwise.

sahil839 commented on web/src/input_pill.js:
```diff
@@ -363,6 +366,21 @@ export function create(opts) {
 
+        // clicking on an user pill should open up the user's profile
+        store.$parent.on("click", ".pill", function (e) {
+            const participant_user_id = Number.parseInt($(this).attr("data-user-id"), 10);
```
It is not clear what is meant by `participant` here. I think having `user_id` as this variable name should be enough.

sahil839 reviewed (commented):
Posted one minor comment. Looks good otherwise. Also, please fix the CI failure.

One thing that I noticed is that the cursor property is set to `not-allowed` for a pill displayed for a person type custom profile field in full user profile modal. But we allow clicking on the pill and it actually opens the popover. Should we change it to the normal cursor now? I think the cursor is set to `not-allowed` because we cannot remove the user from the field and do not show `x` button in the full profile modal.

Pranav2612000 commented:
> Posted one minor comment. Looks good otherwise. Also, please fix the CI failure.
> 
> One thing that I noticed is that the cursor property is set to `not-allowed` for a pill displayed for a person type custom profile field in full user profile modal. But we allow clicking on the pill and it actually opens the popover. Should we change it to the normal cursor now? I think the cursor is set to `not-allowed` because we cannot remove the user from the field and do not show `x` button in the full profile modal.

Yes. Makes sense. I'll do this change as well.

Are there any other places wherer we show a `not-allowed` pill?

Pranav2612000 commented:
@sahil839 Can you take a look again?
I've 
- added some tests to improve coverage and make the CI/CD pass
- removed the `not-allowed` cursor for a better experience
- fixed a bug on the user profile modal where the user popover didn't close on clicking outside

Have broken it down into multiple commits but can squash them later if that's better.

Pranav2612000 commented on web/tests/input_pill.test.js:
```diff
@@ -553,6 +556,79 @@ run_test("exit button on pill", ({mock_template}) => {
     assert.deepEqual(widget.items(), [items.red]);
 });
 
+run_test("open user popover on user's pill click", () => {
```
Are these tests written correctly? It's my first time writing tests at Zulip

sahil839 commented:
I noticed a bug in user profile modal where I come back to the `Profile` tab after going to `Streams` tab once, then I can not open the modal. This is most probably due to the click handler being called twice which closes the modal again. 
![user-pill-in-modal](https://user-images.githubusercontent.com/35494118/223183661-cfc8ae46-cbc9-43b1-a503-7c0e4703fa91.gif)

sahil839 commented on web/src/user_profile.js:
```diff
@@ -376,4 +376,9 @@ export function register_click_handlers() {
         }
     });
+    $("body").on("click", "#user-profile-modal", (e) => {
+        if ($(".user-info-popover").has(e.target).length === 0) {
```
Can we use some function like `popovers.user_info_popped` here?

Pranav2612000 commented on web/src/input_pill.js:
```diff
@@ -363,6 +366,23 @@ export function create(opts) {
         });
 
+        // clicking on an user pill should open up the user's profile
+        store.$parent.off("click", ".pill");
```
@sahil839 You were right. The popup not opening after navigating on streams page was because the click handler was registering twice. The handler gets registered every time a new pill is created.
I'm not sure how we do the clean up in Zulip codebase but I've added this line to ensure that there's only 1 handler at a time which fixes the problem. Let me know your thoughts.

sahil839 replied:
I am not sure whether this is the best solution for this.

The click event attached should ideally be removed when removing the element. But in this case we use a delegated event and not a direct event attached to ".pill" element, and so the click event handler is not removed when the ".pill" element is removed.

sahil839 replied:
One other solution can be to instead use a direct event here, but that should be set after rendering the pill.

I would wait for more feedback on this from Tim when he reviews this. You can keep this as it is for now.

sahil839 reviewed (commented):
Posted one comment about tests.

Pranav2612000 commented:
Cool. @sahil839 I've made the CI/CD passed and addressed your latest commits? Do you think the PR is ready for Tim to review?

sahil839 commented:
I think you missed [this comment](https://github.com/zulip/zulip/pull/24334/commits/6f338c553e58cf66093c7eebd2e4231737fee20f#r1136938349) for tests. There is neither a reply nor any changes in the PR.

Pranav2612000 commented:
> I think you missed [this comment](https://github.com/zulip/zulip/pull/24334/commits/6f338c553e58cf66093c7eebd2e4231737fee20f#r1136938349) for tests. There is neither a reply nor any changes in the PR.

My bad. Thank you for pointing it out. The reason I didn't have the function was because it's not being called. The function which is being tested returns early. I've improved it now by ensuring that the early return works correctly and the other functions are not called. Can you take a look now?

sahil839 commented:
I am not sure how I missed this earlier, but I cannot click on `x` in the user pill to remove a user from user group now since it just opens the popover.

sahil839 commented:
Now, I cannot open the popover for a non-editable pill in user-groups.
![user-groups-pill](https://user-images.githubusercontent.com/35494118/226863728-36820ef3-639b-4c3b-9e9b-87ee26aa423c.gif)

sahil839 commented on web/tests/input_pill.test.js:
```diff
@@ -553,6 +556,90 @@ run_test("exit button on pill", ({mock_template}) => {
+
+    people.get_by_user_id = () => {
+        /* istanbul ignore next */
+        assert.fail("popovers.get_by_user_id should not be called for a non-user pill");
```
There is no need for this, I think. It is enough to test that the popover is not opened.

sahil839 replied:
I don't think we do such checks in other tests, so I don't have any strong opinions. And if someone changes the code to not return early by mistake, then the below code to check `is_user_popover_open` value will anyways fail. We can wait for feedback from others.

Pranav2612000 replied:
Yeah. That's true. I didn't see them anywhere else so I was a bit hesitant. But I still think this is a not a wrong way to test.
> And if someone changes the code to not return early by mistake, then the below code to check is_user_popover_open value will anyways fail.

I think this should be the right behaviour. Anyone modifying the early return code should be conscious of the impact.

Anyway I agree, this may not be the "zulip" convention of doing things.

sahil839 commented on web/tests/input_pill.test.js:
```diff
@@ -553,6 +556,90 @@ run_test("exit button on pill", ({mock_template}) => {
+    /* istanbul ignore next */
+    popovers.show_user_info_popover = () => {
+        is_user_popover_open = true;
+        assert.fail("popovers.show_user_info_popover should not be called for a non-user pill");
```
I don't think we need `assert.fail` statement here, since the test will already fail below where we check `is_user_popover_open` value.

Pranav2612000 replied:
Refer above comment

sahil839 reviewed (commented):
Posted a couple of comments. LGTM otherwise. Also, I think you can squash the last commit with first, since it is necessary to make the clicking of pills work correctly and can mention the reason of the change in commit message.

Pranav2612000 commented:
@sahil839 I've squashed the commits together as requested. The part about `assert.fail` in tests, I've kept it for now, but will remove it as soon as there's another reviewer against it. I hope you don't mind. Are there any other things we need to take care of before asking for a review from others?

alya commented:
Thanks! I'd be OK with this getting fixed as a follow-up if it's complicated, but I noticed that if you click on a person on someone's user profile (e.g., Mentor on dev), and select some option (e.g., View direct messages) from the user card, the original user profile modal remains open. It should probably be closed in this situation, just like we close the settings menu.

alya commented:
I'm not sure how complete my testing was, but looks good to me otherwise.

Pranav2612000 commented:
@alya What would the next steps for this PR be? ( I was thinking of fixing the issue you found as a follow up, unless you think it should go in the same PR? )

Pranav2612000 commented:
@alya Bumping this up^^ 😅

alya commented:
My suggestion was for the fix to go in the same PR if it's easy, and separately if it's not.

alya commented:
@sahil839 Is this PR otherwise ready for integration review from your perspective? If so, please go ahead and add the label.

sahil839 commented:
I don't have any more feedback. @timabbott this is ready for your review. You might want to take a look at tests and the code discussed [here](https://github.com/zulip/zulip/pull/24334#discussion_r1136943746).

Pranav2612000 commented:
@timabbott Can you review this when you have time?

Pranav2612000 commented:
Bumping this up @timabbott @sahil839

sahil839 commented:
@Pranav2612000 Please rebase the PR to resolve merge conflicts and fix the CI as well.

alya commented:
@sahil839 I noticed that https://github.com/zulip/zulip/issues/24247 has been fixed. Is there anything from this PR that's still relevant, or should we close it out?

sahil839 commented:
This PR adds code to open popover by clicking on user pills from anywhere. But there are still some places where we do not open popover on clicking on the pill, like from a user type custom profile field in user profile modal.

Also, the commit for changing the cursor on pills might be useful.

Pranav2612000 commented:
Do we want these changes? If yes, I'll be happy to rebase the PR and bring it up to date. 
If not, feel free to close this.
